# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

In [2]:
source_text[:100].split('\n')

['new jersey is sometimes quiet during autumn , and it is snowy in april .',
 'the united states is usuall']

In [3]:

target_text[:100].split('\n')

["new jersey est parfois calme pendant l' automne , et il est neigeux en avril .",
 'les états-unis est gé']

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [4]:
view_sentence_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 10:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .

French sentences 0 to 10:
new jersey est parfois calme pendant l' automne 

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [5]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    # TODO: Implement Function
    source_text = source_text.split('\n')
    target_text = target_text.split('\n')
    
    source_text_id = []
    for text in source_text:
        source_text_id.append([source_vocab_to_int[i] for i in text.split()])
        
    target_text_id = []
    for text in target_text:
        target_id=[target_vocab_to_int[i] for i in text.split()]
        target_id.append(target_vocab_to_int['<EOS>'])
        target_text_id.append(target_id)
        
    return source_text_id, target_text_id
 

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [6]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [7]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper
import problem_unittests as tests

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [8]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf
from tensorflow.python.layers.core import Dense

# Check TensorFlow Version
assert LooseVersion(tf.__version__) >= LooseVersion('1.1'), 'Please use TensorFlow version 1.1 or newer'
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.2.1
Default GPU Device: /gpu:0


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoder_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.
- Target sequence length placeholder named "target_sequence_length" with rank 1
- Max target sequence length tensor named "max_target_len" getting its value from applying tf.reduce_max on the target_sequence_length placeholder. Rank 0.
- Source sequence length placeholder named "source_sequence_length" with rank 1

Return the placeholders in the following the tuple (input, targets, learning rate, keep probability, target sequence length, max target sequence length, source sequence length)

In [9]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, learning rate, and lengths of source and target sequences.
    :return: Tuple (input, targets, learning rate, keep probability, target sequence length,
    max target sequence length, source sequence length)
    """
    # TODO: Implement Function
    data_input=tf.placeholder(tf.int32,[None,None],name='input')
    targets=tf.placeholder(tf.int32,[None,None],name='targets')
    learning_rate=tf.placeholder(tf.float32,name='learning_rate')
    keep_prob = tf.placeholder(tf.float32, name='keep_prob')
    target_sequence_length = tf.placeholder(tf.int32, [None,], name='target_sequence_length')
    max_target_length = tf.reduce_max(target_sequence_length, name='max_target_length')
    source_sequence_length = tf.placeholder(tf.int32, [None], name='source_sequence_length')
    
    return data_input, targets, learning_rate, keep_prob, target_sequence_length, max_target_length, source_sequence_length


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

ERROR:tensorflow:==================================
Object was never used (type <class 'tensorflow.python.framework.ops.Operation'>):
<tf.Operation 'assert_rank_2/Assert/Assert' type=Assert>
If you want to mark it as used call its "mark_used()" method.
It was originally created here:
['File "/usr/local/lib/python3.5/runpy.py", line 193, in _run_module_as_main\n    "__main__", mod_spec)', 'File "/usr/local/lib/python3.5/runpy.py", line 85, in _run_code\n    exec(code, run_globals)', 'File "/usr/local/lib/python3.5/site-packages/ipykernel_launcher.py", line 16, in <module>\n    app.launch_new_instance()', 'File "/usr/local/lib/python3.5/site-packages/traitlets/config/application.py", line 658, in launch_instance\n    app.start()', 'File "/usr/local/lib/python3.5/site-packages/ipykernel/kernelapp.py", line 477, in start\n    ioloop.IOLoop.instance().start()', 'File "/usr/local/lib/python3.5/site-packages/zmq/eventloop/ioloop.py", line 177, in start\n    super(ZMQIOLoop, self).start()', 'F

Tests Passed


### Process Decoder Input
Implement `process_decoder_input` by removing the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [10]:
def process_decoder_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for encoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    # TODO: Implement Function
    last_word_id = tf.strided_slice(target_data, [0, 0], [batch_size, -1], [1, 1])
    batch = tf.fill([batch_size, 1], target_vocab_to_int['<GO>'])
    concat_batch=tf.concat([batch,last_word_id],1)
    
    return concat_batch

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_encoding_input(process_decoder_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer:
 * Embed the encoder input using [`tf.contrib.layers.embed_sequence`](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embed_sequence)
 * Construct a [stacked](https://github.com/tensorflow/tensorflow/blob/6947f65a374ebf29e74bb71e36fd82760056d82c/tensorflow/docs_src/tutorials/recurrent.md#stacking-multiple-lstms) [`tf.contrib.rnn.LSTMCell`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell) wrapped in a [`tf.contrib.rnn.DropoutWrapper`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/DropoutWrapper)
 * Pass cell and embedded input to [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn)

In [11]:
from imp import reload
reload(tests)

def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   encoding_embedding_size):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :param source_sequence_length: a list of the lengths of each sequence in the batch
    :param source_vocab_size: vocabulary size of source data
    :param encoding_embedding_size: embedding size of source data
    :return: tuple (RNN output, RNN state)
    """
    # TODO: Implement Function
    def cell_make(rnn_size, keep_prob):
        lstm_cell = tf.contrib.rnn.LSTMCell(rnn_size)
        dropout = tf.contrib.rnn.DropoutWrapper(lstm_cell, output_keep_prob=keep_prob)
        return dropout
    
    cell_stack = tf.contrib.rnn.MultiRNNCell([cell_make(rnn_size, keep_prob) for _ in range(num_layers)])
    encoder_embedding = tf.contrib.layers.embed_sequence(rnn_inputs, vocab_size = source_vocab_size, embed_dim = encoding_embedding_size)
    output, state = tf.nn.dynamic_rnn(cell_stack, encoder_embedding, source_sequence_length, dtype=tf.float32)
    return (output, state)
    
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create a training decoding layer:
* Create a [`tf.contrib.seq2seq.TrainingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/TrainingHelper) 
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [12]:

def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                         target_sequence_length, max_summary_length, 
                         output_layer, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_summary_length: The length of the longest sequence in the batch
    :param output_layer: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing training logits and sample_id
    """
    # TODO: Implement Function
    helper_training = tf.contrib.seq2seq.TrainingHelper(inputs=dec_embed_input,sequence_length=target_sequence_length,time_major=False)
    dropout_wrapper = tf.contrib.rnn.DropoutWrapper(dec_cell, output_keep_prob=keep_prob)
    decoder_training = tf.contrib.seq2seq.BasicDecoder(dropout_wrapper, helper_training, encoder_state, output_layer)
    outputs, _, _ = tf.contrib.seq2seq.dynamic_decode(decoder_training, impute_finished=True, maximum_iterations=max_summary_length)
    return outputs

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference decoder:
* Create a [`tf.contrib.seq2seq.GreedyEmbeddingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/GreedyEmbeddingHelper)
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [13]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id,
                         end_of_sequence_id, max_target_sequence_length,
                         vocab_size, output_layer, batch_size, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param max_target_sequence_length: Maximum length of target sequences
    :param vocab_size: Size of decoder/target vocabulary
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_layer: Function to apply the output layer
    :param batch_size: Batch size
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing inference logits and sample_id
    """
    # TODO: Implement Function
    #Reference:https://github.com/udacity/deep-learning/tree/master/seq2seq
    
    start_tokens = tf.tile(tf.constant([start_of_sequence_id], dtype=tf.int32), [batch_size], name='start_tokens')
    helper_infer = tf.contrib.seq2seq.GreedyEmbeddingHelper(dec_embeddings,start_tokens, end_of_sequence_id)
    decoder_infer = tf.contrib.seq2seq.BasicDecoder(dec_cell, helper_infer, encoder_state, output_layer)
    outputs,_,_ = tf.contrib.seq2seq.dynamic_decode(decoder_infer, impute_finished=True,
                                                 maximum_iterations=max_target_sequence_length)    
    return outputs
    

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

* Embed the target sequences
* Construct the decoder LSTM cell (just like you constructed the encoder cell above)
* Create an output layer to map the outputs of the decoder to the elements of our vocabulary
* Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)` function to get the training logits.
* Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [14]:
def decoding_layer(dec_input, encoder_state,
                   target_sequence_length, max_target_sequence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, decoding_embedding_size):
    """
    Create decoding layer
    :param dec_input: Decoder input
    :param encoder_state: Encoder state
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_target_sequence_length: Maximum length of target sequences
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param target_vocab_size: Size of target vocabulary
    :param batch_size: The size of the batch
    :param keep_prob: Dropout keep probability
    :param decoding_embedding_size: Decoding embedding size
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    #Reference:https://github.com/udacity/deep-learning/tree/master/seq2seq
    def cell_make(rnn_size, keep_prob):
        lstm_cell = tf.contrib.rnn.LSTMCell(rnn_size)
        dropout = tf.contrib.rnn.DropoutWrapper(lstm_cell, output_keep_prob=keep_prob)
        return dropout
    
    cell_stack = tf.contrib.rnn.MultiRNNCell([cell_make(rnn_size, keep_prob) for _ in range(num_layers)])
    
    dec_embeddings = tf.Variable(tf.random_uniform([target_vocab_size, decoding_embedding_size]))
    dec_embed_input = tf.nn.embedding_lookup(dec_embeddings, dec_input)
    output = Dense(target_vocab_size,kernel_initializer = tf.truncated_normal_initializer(mean = 0.0, stddev=0.1))
    
    with tf.variable_scope("decode") as scope:
        decoder_output_training = decoding_layer_train(encoder_state, cell_stack, dec_embed_input, target_sequence_length, max_target_sequence_length, output, keep_prob)
        
        scope.reuse_variables()
        
        decoder_output_infer = decoding_layer_infer(encoder_state, cell_stack, dec_embeddings, target_vocab_to_int['<GO>'], target_vocab_to_int['<EOS>'], max_target_sequence_length, target_vocab_size, output, batch_size, keep_prob)
    
    return decoder_output_training, decoder_output_infer


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob,  source_sequence_length, source_vocab_size, encoding_embedding_size)`.
- Process target data using your `process_decoder_input(target_data, target_vocab_to_int, batch_size)` function.
- Decode the encoded input using your `decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)` function.

In [15]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size,
                  source_sequence_length, target_sequence_length,
                  max_target_sentence_length,
                  source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size,
                  rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param source_sequence_length: Sequence Lengths of source sequences in the batch
    :param target_sequence_length: Sequence Lengths of target sequences in the batch
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    #Reference:https://github.com/udacity/deep-learning/tree/master/seq2seq
    _, enc_state = encoding_layer(input_data, rnn_size, num_layers, keep_prob, 
                                           source_sequence_length, source_vocab_size, 
                                           enc_embedding_size)
    dec_input = process_decoder_input(target_data, target_vocab_to_int, batch_size)
    output_train, output_infer  = decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, 
                                                 rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size,
                                                 keep_prob, dec_embedding_size)
    return output_train, output_infer

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability
- Set `display_step` to state how many steps between each debug output statement

In [20]:
# Number of Epochs
epochs = 15
# Batch Size
batch_size = 512
# RNN Size
rnn_size = 128
# Number of Layers
num_layers = 2
# Embedding Size
encoding_embedding_size = 128
decoding_embedding_size = 128
# Learning Rate
learning_rate = 0.001
# Dropout Keep Probability
keep_probability = 0.6
display_step = 5

### Build the Graph
Build the graph using the neural network you implemented.

In [21]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_target_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length = model_inputs()

    #sequence_length = tf.placeholder_with_default(max_target_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)

    train_logits, inference_logits = seq2seq_model(tf.reverse(input_data, [-1]),
                                                   targets,
                                                   keep_prob,
                                                   batch_size,
                                                   source_sequence_length,
                                                   target_sequence_length,
                                                   max_target_sequence_length,
                                                   len(source_vocab_to_int),
                                                   len(target_vocab_to_int),
                                                   encoding_embedding_size,
                                                   decoding_embedding_size,
                                                   rnn_size,
                                                   num_layers,
                                                   target_vocab_to_int)


    training_logits = tf.identity(train_logits.rnn_output, name='logits')
    inference_logits = tf.identity(inference_logits.sample_id, name='predictions')

    masks = tf.sequence_mask(target_sequence_length, max_target_sequence_length, dtype=tf.float32, name='masks')

    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            training_logits,
            targets,
            masks)

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)


Batch and pad the source and target sequences

In [22]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def pad_sentence_batch(sentence_batch, pad_int):
    """Pad sentences with <PAD> so that each sentence of a batch has the same length"""
    max_sentence = max([len(sentence) for sentence in sentence_batch])
    return [sentence + [pad_int] * (max_sentence - len(sentence)) for sentence in sentence_batch]


def get_batches(sources, targets, batch_size, source_pad_int, target_pad_int):
    """Batch targets, sources, and the lengths of their sentences together"""
    for batch_i in range(0, len(sources)//batch_size):
        start_i = batch_i * batch_size

        # Slice the right amount for the batch
        sources_batch = sources[start_i:start_i + batch_size]
        targets_batch = targets[start_i:start_i + batch_size]

        # Pad
        pad_sources_batch = np.array(pad_sentence_batch(sources_batch, source_pad_int))
        pad_targets_batch = np.array(pad_sentence_batch(targets_batch, target_pad_int))

        # Need the lengths for the _lengths parameters
        pad_targets_lengths = []
        for target in pad_targets_batch:
            pad_targets_lengths.append(len(target))

        pad_source_lengths = []
        for source in pad_sources_batch:
            pad_source_lengths.append(len(source))

        yield pad_sources_batch, pad_targets_batch, pad_source_lengths, pad_targets_lengths


### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [23]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1])],
            'constant')

    return np.mean(np.equal(target, logits))

# Split data to training and validation sets
train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]
valid_source = source_int_text[:batch_size]
valid_target = target_int_text[:batch_size]
(valid_sources_batch, valid_targets_batch, valid_sources_lengths, valid_targets_lengths ) = next(get_batches(valid_source,
                                                                                                             valid_target,
                                                                                                             batch_size,
                                                                                                             source_vocab_to_int['<PAD>'],
                                                                                                             target_vocab_to_int['<PAD>']))                                                                                                  
with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch, sources_lengths, targets_lengths) in enumerate(
                get_batches(train_source, train_target, batch_size,
                            source_vocab_to_int['<PAD>'],
                            target_vocab_to_int['<PAD>'])):

            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 target_sequence_length: targets_lengths,
                 source_sequence_length: sources_lengths,
                 keep_prob: keep_probability})


            if batch_i % display_step == 0 and batch_i > 0:


                batch_train_logits = sess.run(
                    inference_logits,
                    {input_data: source_batch,
                     source_sequence_length: sources_lengths,
                     target_sequence_length: targets_lengths,
                     keep_prob: 1.0})


                batch_valid_logits = sess.run(
                    inference_logits,
                    {input_data: valid_sources_batch,
                     source_sequence_length: valid_sources_lengths,
                     target_sequence_length: valid_targets_lengths,
                     keep_prob: 1.0})

                train_acc = get_accuracy(target_batch, batch_train_logits)

                valid_acc = get_accuracy(valid_targets_batch, batch_valid_logits)

                print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.4f}, Validation Accuracy: {:>6.4f}, Loss: {:>6.4f}'
                      .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch    5/269 - Train Accuracy: 0.2325, Validation Accuracy: 0.3096, Loss: 4.9724
Epoch   0 Batch   10/269 - Train Accuracy: 0.2567, Validation Accuracy: 0.3314, Loss: 4.5245
Epoch   0 Batch   15/269 - Train Accuracy: 0.2960, Validation Accuracy: 0.3428, Loss: 4.0512
Epoch   0 Batch   20/269 - Train Accuracy: 0.3086, Validation Accuracy: 0.3715, Loss: 3.8318
Epoch   0 Batch   25/269 - Train Accuracy: 0.3192, Validation Accuracy: 0.3811, Loss: 3.6277
Epoch   0 Batch   30/269 - Train Accuracy: 0.3742, Validation Accuracy: 0.4082, Loss: 3.3438
Epoch   0 Batch   35/269 - Train Accuracy: 0.3835, Validation Accuracy: 0.4127, Loss: 3.1821
Epoch   0 Batch   40/269 - Train Accuracy: 0.3559, Validation Accuracy: 0.4179, Loss: 3.2125
Epoch   0 Batch   45/269 - Train Accuracy: 0.3765, Validation Accuracy: 0.4356, Loss: 3.0975
Epoch   0 Batch   50/269 - Train Accuracy: 0.3954, Validation Accuracy: 0.4477, Loss: 3.0290
Epoch   0 Batch   55/269 - Train Accuracy: 0.4228, Validation Accuracy

Epoch   1 Batch  185/269 - Train Accuracy: 0.5517, Validation Accuracy: 0.5656, Loss: 0.9858
Epoch   1 Batch  190/269 - Train Accuracy: 0.5539, Validation Accuracy: 0.5662, Loss: 0.9741
Epoch   1 Batch  195/269 - Train Accuracy: 0.5326, Validation Accuracy: 0.5634, Loss: 0.9933
Epoch   1 Batch  200/269 - Train Accuracy: 0.5336, Validation Accuracy: 0.5682, Loss: 1.0095
Epoch   1 Batch  205/269 - Train Accuracy: 0.5376, Validation Accuracy: 0.5689, Loss: 0.9674
Epoch   1 Batch  210/269 - Train Accuracy: 0.5607, Validation Accuracy: 0.5649, Loss: 0.9585
Epoch   1 Batch  215/269 - Train Accuracy: 0.5799, Validation Accuracy: 0.5724, Loss: 0.8917
Epoch   1 Batch  220/269 - Train Accuracy: 0.5620, Validation Accuracy: 0.5650, Loss: 0.9088
Epoch   1 Batch  225/269 - Train Accuracy: 0.5409, Validation Accuracy: 0.5673, Loss: 0.9716
Epoch   1 Batch  230/269 - Train Accuracy: 0.5514, Validation Accuracy: 0.5655, Loss: 0.9381
Epoch   1 Batch  235/269 - Train Accuracy: 0.5511, Validation Accuracy

Epoch   3 Batch  100/269 - Train Accuracy: 0.6457, Validation Accuracy: 0.6282, Loss: 0.6668
Epoch   3 Batch  105/269 - Train Accuracy: 0.6145, Validation Accuracy: 0.6342, Loss: 0.6868
Epoch   3 Batch  110/269 - Train Accuracy: 0.6192, Validation Accuracy: 0.6303, Loss: 0.6540
Epoch   3 Batch  115/269 - Train Accuracy: 0.6156, Validation Accuracy: 0.6340, Loss: 0.6879
Epoch   3 Batch  120/269 - Train Accuracy: 0.6178, Validation Accuracy: 0.6284, Loss: 0.6835
Epoch   3 Batch  125/269 - Train Accuracy: 0.6317, Validation Accuracy: 0.6345, Loss: 0.6392
Epoch   3 Batch  130/269 - Train Accuracy: 0.6126, Validation Accuracy: 0.6423, Loss: 0.6888
Epoch   3 Batch  135/269 - Train Accuracy: 0.5983, Validation Accuracy: 0.6328, Loss: 0.7126
Epoch   3 Batch  140/269 - Train Accuracy: 0.6283, Validation Accuracy: 0.6307, Loss: 0.6738
Epoch   3 Batch  145/269 - Train Accuracy: 0.6370, Validation Accuracy: 0.6358, Loss: 0.6547
Epoch   3 Batch  150/269 - Train Accuracy: 0.6331, Validation Accuracy

Epoch   5 Batch   15/269 - Train Accuracy: 0.6791, Validation Accuracy: 0.6890, Loss: 0.4906
Epoch   5 Batch   20/269 - Train Accuracy: 0.6826, Validation Accuracy: 0.7008, Loss: 0.5226
Epoch   5 Batch   25/269 - Train Accuracy: 0.6854, Validation Accuracy: 0.6959, Loss: 0.5423
Epoch   5 Batch   30/269 - Train Accuracy: 0.6997, Validation Accuracy: 0.6984, Loss: 0.4919
Epoch   5 Batch   35/269 - Train Accuracy: 0.6938, Validation Accuracy: 0.7001, Loss: 0.5113
Epoch   5 Batch   40/269 - Train Accuracy: 0.6839, Validation Accuracy: 0.7088, Loss: 0.5148
Epoch   5 Batch   45/269 - Train Accuracy: 0.6937, Validation Accuracy: 0.7051, Loss: 0.5155
Epoch   5 Batch   50/269 - Train Accuracy: 0.6796, Validation Accuracy: 0.7011, Loss: 0.5187
Epoch   5 Batch   55/269 - Train Accuracy: 0.7191, Validation Accuracy: 0.6998, Loss: 0.4854
Epoch   5 Batch   60/269 - Train Accuracy: 0.7039, Validation Accuracy: 0.6970, Loss: 0.4758
Epoch   5 Batch   65/269 - Train Accuracy: 0.6749, Validation Accuracy

Epoch   6 Batch  195/269 - Train Accuracy: 0.7309, Validation Accuracy: 0.7504, Loss: 0.3863
Epoch   6 Batch  200/269 - Train Accuracy: 0.7570, Validation Accuracy: 0.7740, Loss: 0.3966
Epoch   6 Batch  205/269 - Train Accuracy: 0.7786, Validation Accuracy: 0.7681, Loss: 0.3790
Epoch   6 Batch  210/269 - Train Accuracy: 0.7729, Validation Accuracy: 0.7684, Loss: 0.3700
Epoch   6 Batch  215/269 - Train Accuracy: 0.7876, Validation Accuracy: 0.7527, Loss: 0.3599
Epoch   6 Batch  220/269 - Train Accuracy: 0.7819, Validation Accuracy: 0.7728, Loss: 0.3648
Epoch   6 Batch  225/269 - Train Accuracy: 0.7668, Validation Accuracy: 0.7588, Loss: 0.3771
Epoch   6 Batch  230/269 - Train Accuracy: 0.7790, Validation Accuracy: 0.7755, Loss: 0.3700
Epoch   6 Batch  235/269 - Train Accuracy: 0.7758, Validation Accuracy: 0.7666, Loss: 0.3649
Epoch   6 Batch  240/269 - Train Accuracy: 0.7961, Validation Accuracy: 0.7707, Loss: 0.3459
Epoch   6 Batch  245/269 - Train Accuracy: 0.7566, Validation Accuracy

Epoch   8 Batch  110/269 - Train Accuracy: 0.8232, Validation Accuracy: 0.8263, Loss: 0.2900
Epoch   8 Batch  115/269 - Train Accuracy: 0.8149, Validation Accuracy: 0.8347, Loss: 0.3166
Epoch   8 Batch  120/269 - Train Accuracy: 0.8199, Validation Accuracy: 0.8246, Loss: 0.3107
Epoch   8 Batch  125/269 - Train Accuracy: 0.8481, Validation Accuracy: 0.8260, Loss: 0.2826
Epoch   8 Batch  130/269 - Train Accuracy: 0.8097, Validation Accuracy: 0.8224, Loss: 0.3229
Epoch   8 Batch  135/269 - Train Accuracy: 0.8009, Validation Accuracy: 0.8090, Loss: 0.3212
Epoch   8 Batch  140/269 - Train Accuracy: 0.8238, Validation Accuracy: 0.8143, Loss: 0.3131
Epoch   8 Batch  145/269 - Train Accuracy: 0.8228, Validation Accuracy: 0.8232, Loss: 0.2844
Epoch   8 Batch  150/269 - Train Accuracy: 0.8374, Validation Accuracy: 0.8367, Loss: 0.2958
Epoch   8 Batch  155/269 - Train Accuracy: 0.8297, Validation Accuracy: 0.8264, Loss: 0.2814
Epoch   8 Batch  160/269 - Train Accuracy: 0.8344, Validation Accuracy

Epoch  10 Batch   25/269 - Train Accuracy: 0.8563, Validation Accuracy: 0.8729, Loss: 0.2617
Epoch  10 Batch   30/269 - Train Accuracy: 0.8561, Validation Accuracy: 0.8644, Loss: 0.2389
Epoch  10 Batch   35/269 - Train Accuracy: 0.8691, Validation Accuracy: 0.8746, Loss: 0.2508
Epoch  10 Batch   40/269 - Train Accuracy: 0.8590, Validation Accuracy: 0.8681, Loss: 0.2533
Epoch  10 Batch   45/269 - Train Accuracy: 0.8696, Validation Accuracy: 0.8643, Loss: 0.2420
Epoch  10 Batch   50/269 - Train Accuracy: 0.8329, Validation Accuracy: 0.8639, Loss: 0.2501
Epoch  10 Batch   55/269 - Train Accuracy: 0.8837, Validation Accuracy: 0.8693, Loss: 0.2269
Epoch  10 Batch   60/269 - Train Accuracy: 0.8595, Validation Accuracy: 0.8616, Loss: 0.2287
Epoch  10 Batch   65/269 - Train Accuracy: 0.8586, Validation Accuracy: 0.8736, Loss: 0.2337
Epoch  10 Batch   70/269 - Train Accuracy: 0.8688, Validation Accuracy: 0.8658, Loss: 0.2416
Epoch  10 Batch   75/269 - Train Accuracy: 0.8813, Validation Accuracy

Epoch  11 Batch  205/269 - Train Accuracy: 0.8663, Validation Accuracy: 0.8736, Loss: 0.2030
Epoch  11 Batch  210/269 - Train Accuracy: 0.8840, Validation Accuracy: 0.8875, Loss: 0.2000
Epoch  11 Batch  215/269 - Train Accuracy: 0.8999, Validation Accuracy: 0.8883, Loss: 0.1929
Epoch  11 Batch  220/269 - Train Accuracy: 0.8856, Validation Accuracy: 0.8888, Loss: 0.1974
Epoch  11 Batch  225/269 - Train Accuracy: 0.8887, Validation Accuracy: 0.8936, Loss: 0.1910
Epoch  11 Batch  230/269 - Train Accuracy: 0.8801, Validation Accuracy: 0.8884, Loss: 0.1983
Epoch  11 Batch  235/269 - Train Accuracy: 0.8847, Validation Accuracy: 0.8836, Loss: 0.2004
Epoch  11 Batch  240/269 - Train Accuracy: 0.9027, Validation Accuracy: 0.8908, Loss: 0.1892
Epoch  11 Batch  245/269 - Train Accuracy: 0.8781, Validation Accuracy: 0.8924, Loss: 0.2153
Epoch  11 Batch  250/269 - Train Accuracy: 0.8961, Validation Accuracy: 0.8874, Loss: 0.1992
Epoch  11 Batch  255/269 - Train Accuracy: 0.8870, Validation Accuracy

Epoch  13 Batch  120/269 - Train Accuracy: 0.9011, Validation Accuracy: 0.9017, Loss: 0.1738
Epoch  13 Batch  125/269 - Train Accuracy: 0.9043, Validation Accuracy: 0.8983, Loss: 0.1663
Epoch  13 Batch  130/269 - Train Accuracy: 0.9057, Validation Accuracy: 0.8991, Loss: 0.1780
Epoch  13 Batch  135/269 - Train Accuracy: 0.8982, Validation Accuracy: 0.8964, Loss: 0.1827
Epoch  13 Batch  140/269 - Train Accuracy: 0.8947, Validation Accuracy: 0.9048, Loss: 0.1964
Epoch  13 Batch  145/269 - Train Accuracy: 0.9002, Validation Accuracy: 0.9060, Loss: 0.1679
Epoch  13 Batch  150/269 - Train Accuracy: 0.8958, Validation Accuracy: 0.8996, Loss: 0.1632
Epoch  13 Batch  155/269 - Train Accuracy: 0.9026, Validation Accuracy: 0.9088, Loss: 0.1626
Epoch  13 Batch  160/269 - Train Accuracy: 0.8998, Validation Accuracy: 0.9102, Loss: 0.1643
Epoch  13 Batch  165/269 - Train Accuracy: 0.9143, Validation Accuracy: 0.9055, Loss: 0.1655
Epoch  13 Batch  170/269 - Train Accuracy: 0.9051, Validation Accuracy

### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [24]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [25]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [26]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    # TODO: Implement Function
    words_to_id=[]
    lower_sentence=sentence.lower().split()
    for i in lower_sentence:
        
        if i in vocab_to_int:
            words_to_id.append(vocab_to_int[i])
            
        else:
            words_to_id.append(vocab_to_int['<UNK>'])
    return words_to_id


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


## Translate
This will translate `translate_sentence` from English to French.

In [27]:
translate_sentence = 'he saw a old yellow truck .'


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('predictions:0')
    target_sequence_length = loaded_graph.get_tensor_by_name('target_sequence_length:0')
    source_sequence_length = loaded_graph.get_tensor_by_name('source_sequence_length:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence]*batch_size,
                                         target_sequence_length: [len(translate_sentence)*2]*batch_size,
                                         source_sequence_length: [len(translate_sentence)]*batch_size,
                                         keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in translate_logits]))
print('  French Words: {}'.format(" ".join([target_int_to_vocab[i] for i in translate_logits])))


INFO:tensorflow:Restoring parameters from checkpoints/dev
Input
  Word Ids:      [124, 118, 15, 30, 91, 58, 158]
  English Words: ['he', 'saw', 'a', 'old', 'yellow', 'truck', '.']

Prediction
  Word Ids:      [161, 104, 310, 71, 263, 117, 218, 73, 1]
  French Words: il a vu un vieux camion jaune . <EOS>


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.